eZ/Publish/Core/FieldType/Keyword/KeywordStorage/Gateway/LegacyStorage.php

```diff
@@ -281,9 +287,17 @@ protected function insertKeywords(array $keywordsToInsert, $contentTypeId)
         return $keywordIdMap;
     }
 
-    protected function deleteOldKeywordAssignments($fieldId)
+    protected function deleteOldKeywordAssignments($fieldId, VersionInfo $versionInfo = null)
     {
         $dbHandler = $this->getConnection();
+        // If current version being asked to be deleted is not published, then don't do that if
+        // there is some other version which is published (as keyword table is not versioned)
+        if ($versionInfo !== null &&
+            $versionInfo->status !== VersionInfo::STATUS_PUBLISHED &&
+            $versionInfo->contentInfo->isPublished
+        ) {
+            return;
+        }
```

Comment:
Not sure if Gateway is a right place to perform this check. I would move this check to `KeywordStorage` before `foreach` and prevent calls to gateway that silently do nothing. 

Ref. https://github.com/ezsystems/ezpublish-kernel/pull/1923/files#r105897607